src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12912,12 +12912,32 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
         private PostSkipAction SkipBadInitializerListTokens<T>(ref SyntaxToken startToken, SeparatedSyntaxListBuilder<T> list, SyntaxKind expected)
             where T : CSharpSyntaxNode
         {
+            bool openBraceIsMissing = startToken.IsMissing;
             return this.SkipBadSeparatedListTokensWithExpectedKind(ref startToken, list,
                 p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleExpression(),
-                p => p.CurrentToken.Kind == SyntaxKind.CloseBraceToken || p.IsTerminator(),
+                p => p.SkipBadInitializerToken(openBraceIsMissing),
                 expected);
         }
 
+        private bool SkipBadInitializerToken(bool openBraceIsMissing)
+        {
+            if (!openBraceIsMissing && CurrentToken.Kind == SyntaxKind.SemicolonToken)
+            {
+                return !isNextToken(lp => lp.CurrentToken.Kind == SyntaxKind.CloseBraceToken || lp.IsPossibleVariableInitializer());
+            }
+            return CurrentToken.Kind == SyntaxKind.CloseBraceToken || IsTerminator();
+
+            bool isNextToken(Func<LanguageParser, bool> nextTokenParseFunc)
+            {
+                var resetPoint = GetResetPoint();
+                EatToken();
+                bool result = nextTokenParseFunc(this);
```

Comment:
i don't see the benefit in passing a callback here.  we know exactly the code to run to check this.